allow started tasks to be tracked

tbg opened the pull request:
it's interesting to see which tasks are running, especially
when graceful shutdown does not work (see 1490).

```
[...]server/cli/start.go:166] 3 running tasks
    1x /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/storage/queue.go:238
    2x /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:189
```

the `StartTask` api needed to change for that. I'm happy with the new one,
but that may be my personal preference and if it doesn't sit well with the rest
of you, I'm ok with that.

tamird commented on server/node.go:
```diff
@@ -309,11 +309,14 @@ func (n *Node) initStores(engines []engine.Engine, stopper *util.Stopper) error
-			stopper.FinishTask()
-		}()
+	if bootstraps.Len() > 0 {
+		task := stopper.StartTask()
```
`if task := ... ; task.Ok()`

tschottdorf replied:
done

tamird commented on util/stopper.go:
```diff
@@ -88,40 +92,83 @@ func (s *Stopper) AddCloser(c Closer) {
+// Done is called (exactly once) when the started work has completed.
+// It does not need to be called when Ok() returned false.
+func (pt PendingTask) Done() {
+	if pt.Ok() {
```
any reason not to panic here?

tschottdorf replied:
no. Added.

tamird commented on util/stopper.go:
```diff
@@ -88,40 +92,83 @@ func (s *Stopper) AddCloser(c Closer) {
 
-// NumTasks returns the number of active tasks.
-func (s *Stopper) NumTasks() int {
+// RunningTasks returns a map containing the numbers of running maps
```
running tasks

tschottdorf replied:
done.

tamird commented on util/stopper.go:
```diff
@@ -88,40 +92,83 @@ func (s *Stopper) AddCloser(c Closer) {
+	var finished bool
+	return func() bool {
+		if finished {
+			panic("cannot finish a task twice")
```
instead of doing this here, why not have Done() set the function pointer back to null?

tschottdorf replied:
now that I'm panicking on `Done()` when the task wasn't allowed, I'll keep it to have different panics for the two cases.

tamird commented on util/stopper_test.go:
```diff
@@ -237,3 +236,52 @@ func TestStopperClosers(t *testing.T) {
+		}
+		m := s.RunningTasks()
+		if len(m) != 1 {
+			t.Fatalf("expected exactly one task map entry: %+v", m)
```
nit: why fatal?

mrtracy replied:
I think its better to prefer Fatal over Error in order to speed up test runs, unless you're certain that continuing the test may yield more useful information.  For instance, if there are several orthagonal test cases being run in a loop, or if you perform several assertions on a complicated return value, it's probably appropriate to use Errorf.

However, these assertions appear to indicate that test setup has failed in some way; there's no way the rest of the test could be relied upon, so its appropriate to fail fast.

tschottdorf replied:
it's all fairly incremental and if it fails, I usually don't want to sift through the log finding out where it all began.

bdarnell replied:
I prefer to use `t.Fatal` by default and only use `t.Error` when there is a reasonable expectation that collecting multiple errors will be informative. The big reason is that if there is a panic (which is fairly likely if the error resulted in some pointer being set to nil), the test process is simply killed (with a stack trace at the point of the panic) and the errors that have been collected by `t.Error` are never printed.

tamird commented on util/stopper_test.go:
```diff
@@ -237,3 +236,52 @@ func TestStopperClosers(t *testing.T) {
+		}
+		for _, v := range m {
+			if v != i+1 {
+				t.Fatalf("expected %d tasks, got %d", i+1, v)
```
ditto

tschottdorf replied:
ditto

tamird commented on util/stopper_test.go:
```diff
@@ -237,3 +236,52 @@ func TestStopperClosers(t *testing.T) {
+
+	var i int
+	for _, task := range tasks {
+		i++
```
`for i, task := range tasks` would be equivalent, no?

tschottdorf replied:
yes mod off by one, must've thought `tasks` was a map.

tamird commented on util/stopper_test.go:
```diff
@@ -237,3 +236,52 @@ func TestStopperClosers(t *testing.T) {
+			t.Fatalf("task should have been allowed")
+		}
+		task.Done()
+		if numTypes, numTasks, expNum := len(s.tasks), s.NumTasks(), 3-i; numTypes != 1 || numTasks != expNum {
```
this `3-i` is really len(tasks[i:]) or len(tasks[i+1:]) if `i` is zero indexed as i suggested above

tschottdorf replied:
done.

tamird commented on util/stopper_test.go:
```diff
@@ -237,3 +236,52 @@ func TestStopperClosers(t *testing.T) {
+		}
+		task.Done()
+		if numTypes, numTasks, expNum := len(s.tasks), s.NumTasks(), 3-i; numTypes != 1 || numTasks != expNum {
+			t.Errorf("stopper should have %d running tasks, got %d / %+v", 3-i, s.NumTasks(), s.tasks)
```
s/3-i/expNum

tschottdorf replied:
done.

tamird commented on util/stopper_test.go:
```diff
@@ -237,3 +236,52 @@ func TestStopperClosers(t *testing.T) {
+		}
+		m := s.RunningTasks()
+		// Done() on the last task should clear out the map.
+		if len(m) != 1 && i%3 != 0 {
```
maybe do this at the top of the loop and then again outside the loop for the empty case?

tamird replied:
oh you already have the outside check - just move this to the top of the loop

tschottdorf replied:
done.

tamird commented on util/stopper_test.go:
```diff
@@ -237,3 +236,52 @@ func TestStopperClosers(t *testing.T) {
+	}
+
+	s.Stop()
+
```
nit: trailing newline

tschottdorf replied:
done.

tamird commented on util/stopper_test.go:
```diff
@@ -237,3 +236,52 @@ func TestStopperClosers(t *testing.T) {
+		t.Fatalf("task map not empty: %+v", m)
+	}
+
+	s.Stop()
```
nit: defer?

tschottdorf replied:
done.

vivekmenezes commented on util/stopper.go:
```diff
@@ -88,40 +92,83 @@ func (s *Stopper) AddCloser(c Closer) {
+	_, file, line, _ := runtime.Caller(1)
+	taskKey := fmt.Sprintf("%s:%d", file, line)
+	s.tasks[taskKey]++
+	var finished bool
```
s/finished/done to match Done()

tschottdorf replied:
done.

vivekmenezes commented on util/stopper.go:
```diff
@@ -88,40 +92,83 @@ func (s *Stopper) AddCloser(c Closer) {
+// Ok() must be called to confirm that the task can be launched. Ok() returns
+// false to indicate that the system is currently draining and the task should
+// must not proceed. It's not necessary to call Done() in that case.
+func (s *Stopper) StartTask() PendingTask {
```
while the functionality is fine here. I feel that this abstraction matches more one of a work permit.

I prefer

permit := stopper.GetPermit() WorkPermit {

if (permit.Issued()) {
  defer permit.Return()
  // do work
}
If you like that, I'd use gorename

token might be another option, but task doesn't seem right, because it's not a task.

petermattis replied:
Might make sense, but let's leave the naming along in this PR.

mrtracy replied:
+1 for permit

tschottdorf replied:
added the comments here and @bdarnell's into a TODO.

petermattis commented on util/stopper.go:
```diff
@@ -88,40 +92,83 @@ func (s *Stopper) AddCloser(c Closer) {
-// First movers include goroutines launched to do periodic work and
-// the kv/db.go gateway which accepts external client
-// requests.
+// A PendingTask is returned by StartTask. For simplicity this is just a
```
`s/PendingTask/RunningTask/g`, because the task is running, correct?

tschottdorf replied:
that's the thing, it may not be. You need to call `Ok()` first. There's a TODO about the interface now.

petermattis commented on util/stopper.go:
```diff
@@ -88,40 +92,83 @@ func (s *Stopper) AddCloser(c Closer) {
-func (s *Stopper) NumTasks() int {
+// RunningTasks returns a map containing the numbers of running maps
+// keyed by callsite.
+func (s *Stopper) RunningTasks() map[string]int {
```
Do you ever need the map? Would it be sufficient to return the formatted string?

tschottdorf replied:
wanna leave it to the caller. Gave this an own `fmt.Stringer` though.

petermattis commented:
LGTM

mrtracy commented on util/stopper_test.go:
```diff
@@ -237,3 +236,52 @@ func TestStopperClosers(t *testing.T) {
+	for i := 0; i < 3; i++ {
+		tasks = append(tasks, s.StartTask())
+		if numTypes, numTasks := len(s.tasks), s.NumTasks(); numTypes != 1 || numTasks != i+1 {
+			t.Errorf("stopper should have %d running tasks, got %d / %+v", i+1, s.NumTasks(), s.tasks)
```
Use the `numTypes` and `numTasks` in the format string, instead of computing the values again.

tschottdorf replied:
done.

bdarnell commented:
LGTM, but the client interface is getting cumbersome. How about a `RunTask` method, similar to `RunWorker` which hid `AddWorker` and `SetStopped`? Then the `task` object can be local to the stopper instead of being exposed to the client.

vivekmenezes commented on kv/txn_coord_sender.go:
```diff
@@ -196,7 +196,7 @@ func (tm *txnMetadata) close(trace *tracer.Trace, txn *proto.Transaction, resolv
 					log.Warningf("failed to cleanup %q intent: %s", call.Args.Header().Key, call.Reply.Header().GoError())
 				}
-				stopper.FinishTask()
+				task.Done()
```
perhaps you can move this up to the top and add a defer

tschottdorf replied:
done.

vivekmenezes commented:
Looks okay, but instead of the "permit" idea I suggested I think it would be nicer to just pass a function to StartTask to execute if the stopper is not draining. You can check this in and I could do that as an exercise.

tschottdorf commented:
yes, that's @bdarnell's comment and a part of the TODO. Feel free to grab this when it merges.

tbg committed:
allow started tasks to be tracked

it's interesting to see which tasks are running, especially
when graceful shutdown does not work (see 1490).

```
[...]server/cli/start.go:166] 3 running tasks
	1x /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/storage/queue.go:238
	2x /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:189
```

tschottdorf commented:
btw, I'm worried that that'll lead to a bunch of awkward function closures. The nice thing about StartTask as it was is the ability to exit early and then proceed in the main flow. But only one way to find out.
